weave/integrations/langchain/langchain_test.py

```diff
@@ -0,0 +1,494 @@
+        "Question: {question}\nContext: {context}\nAnswer:"
+    )
+
+    llm = ChatOpenAI(
+        model_name="gpt-3.5-turbo", openai_api_key=api_key, temperature=0.0
+    )
+
+    def format_docs(documents: List[Document]) -> str:
+        return "\n\n".join(doc.page_content for doc in documents)
+
+    # Chain
+    rag_chain = (
+        {"context": retriever | format_docs, "question": RunnablePassthrough()}
+        | prompt
+        | llm
+        | StrOutputParser()
+    )
+
+    rag_chain.invoke(
+        input="What is the essay about?",
+    )
+
+    res = client.server.calls_query(tsi.CallsQueryReq(project_id=client._project_id()))
+    assert_correct_calls_for_rag_chain(res.calls)
+
+
+def assert_correct_calls_for_agent_with_tool(calls: list[tsi.CallSchema]) -> None:
+    assert len(calls) == 10
+
+    flattened = flatten_calls(calls)
+
+    got = [(op_name_from_ref(c.op_name), d) for (c, d) in flattened]
+
+    exp = [
+        ("langchain.Chain.AgentExecutor", 0),
+        ("langchain.Chain.RunnableSequence", 1),
+        ("langchain.Chain.RunnableParallel input chat_history agent_scratchpad ", 2),
+        ("langchain.Chain.RunnableLambda", 3),
+        ("langchain.Chain.RunnableLambda", 3),
+        ("langchain.Chain.RunnableLambda", 3),
+        ("langchain.Prompt.ChatPromptTemplate", 2),
+        ("langchain.Llm.ChatOpenAI", 2),
+        ("openai.chat.completions.create", 3),
+        ("langchain.Parser.OpenAIFunctionsAgentOutputParser", 2),
+    ]
+    assert got == exp
+
+
+@pytest.mark.skip_clickhouse_client
+@pytest.mark.vcr(
+    filter_headers=["authorization"],
+    allowed_hosts=["api.wandb.ai", "localhost", "trace.wandb.ai"],
+    before_record_request=filter_body,
+)
+def test_agent_run_with_tools(
+    client: WeaveClient,
+) -> None:
+    from langchain.agents import AgentExecutor
+    from langchain.agents.format_scratchpad import format_to_openai_function_messages
+    from langchain.agents.output_parsers import OpenAIFunctionsAgentOutputParser
+    from langchain.pydantic_v1 import BaseModel, Field
+    from langchain.tools import StructuredTool
+    from langchain_core.messages import AIMessage, HumanMessage
+    from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
+    from langchain_core.utils.function_calling import convert_to_openai_tool
+    from langchain_openai import ChatOpenAI
+
+    class CalculatorInput(BaseModel):
+        a: int = Field(description="first number")
+        b: int = Field(description="second number")
+
+    def multiply(a: int, b: int) -> int:
+        """Multiply two numbers."""
+        return a * b
+
+    calculator = StructuredTool.from_function(
+        func=multiply,
+        name="Calculator",
+        description="multiply numbers",
+        args_schema=CalculatorInput,
+        return_direct=True,
+    )
+
+    api_key = os.environ.get("OPENAI_API_KEY", "sk-1234567890abcdef1234567890abcdef")
```

Comment:
api key.

Reply:
Haha,  :laughing: this is a dummy key made to look the same cosmetically